File: src/code_indexing_mcp/installer/uninstall.py
"""Taking back exactly what the installer put down.

The rule throughout: remove only what this installer created, and recognise it
by evidence rather than by name. A file at the launcher's path that is not one
of our symlinks, a skill directory entry that is not one of our links, a PATH
block whose end marker the user deleted -- each stays where it is, and the
uninstaller says so instead of guessing.
"""

from __future__ import annotations

import os
import shutil
from collections.abc import Callable, Mapping
from dataclasses import dataclass, field
from pathlib import Path

from . import harnesses, shell_path
from .config_files import InstallerError
from .links import is_under
from .orchestrator import StepEvent, default_install_directory


@dataclass(frozen=True)
class UninstallPlan:
    install_directory: Path
    harness_slugs: tuple[str, ...] = ()
    bin_directory: Path | None = None
    remove_launcher: bool = True
    remove_path_block: bool = True
    # Off by default: indexes cost minutes of CPU to rebuild, and an uninstall
    # that silently discards them is not one the user can undo.
    remove_data: bool = False
    # Off by default: the checkout may be a working copy the user edits.
    remove_checkout: bool = False


@dataclass
class UninstallResult:
    harnesses_cleared: tuple[tuple[str, Path, bool], ...] = ()
    skills: tuple[tuple[str, str], ...] = ()
    launcher_removed: Path | None = None
    profiles_cleared: tuple[Path, ...] = ()
    directories_removed: tuple[Path, ...] = ()
    failures: list[tuple[str, str]] = field(default_factory=list)


def data_directories(
    *,
    environment: Mapping[str, str] | None = None,
) -> tuple[Path, ...]:
    """The index and cache directories, honouring the settings that move them."""

    from .settings_spec import BY_NAME, default_value

    environment = os.environ if environment is None else environment
    directories: list[Path] = []
    for name in ("CODE_INDEXING_DATA_DIR", "CODE_INDEXING_CACHE_DIR"):
        setting = BY_NAME.get(name)
        if setting is None:  # pragma: no cover - the catalog defines both
            continue
        configured = environment.get(name) or default_value(setting)
        if configured:
            directories.append(Path(configured).expanduser())
    return tuple(directories)


# Files and directories the server itself creates under its data or cache
# directory. One of these present is what distinguishes "our index lives here"
# from "the user pointed the setting at a directory that holds other things".
_DATA_MARKERS = (
    "lancedb",
    "locks",
    "staging",
    "accelerator.json",
    "daemon.token",
    "daemon.log",
    "models",
)


def _refuse_reason(directory: Path, *, checkout: bool, home: Path | None = None) -> str | None:
    """Why ``directory`` must not be deleted, or None when removing it is safe.

    ``--purge`` and ``--remove-checkout`` take their targets from a setting and a
    flag, either of which can name somewhere that is not ours. The confirmation
    prompt shows what will go, but a prompt is not a safety net for a recursive
    delete that cannot be undone: the directory has to look like ours as well.
    """

    home = home or Path.home()
    try:
        resolved = directory.resolve()
    except (OSError, ValueError) as exc:
        return f"cannot be resolved ({exc})"
    if resolved.parent == resolved:
        return "is a filesystem root"
    if resolved == home.resolve():
        return "is your home directory"
    if is_under(home.resolve(), resolved):
        return "contains your home directory"
    if checkout:
        if not (resolved / "pyproject.toml").is_file():
            return "does not look like a code-indexing-mcp checkout (no pyproject.toml)"
        if not (resolved / "src" / "code_indexing_mcp").is_dir():
            return "does not look like a code-indexing-mcp checkout (no src/code_indexing_mcp)"
        return None
    if resolved.name == "code-indexing-mcp":
        return None
    if any((resolved / marker).exists() for marker in _DATA_MARKERS):
        return None
    return "holds no code-indexing-mcp index or cache, so it is not ours to delete"


def run_uninstall(
    plan: UninstallPlan,
    on_event: Callable[[StepEvent], None] = lambda event: None,
    *,
    home: Path | None = None,
    environment: Mapping[str, str] | None = None,
) -> UninstallResult:
    """Undo an installation, step by step, reporting each removal.

    Every step isolates its own failures: a config file that cannot be written
    must not stop the launcher and the PATH entry from being taken back.
    """

    result = UninstallResult()

    on_event(StepEvent("harnesses", "started", ", ".join(plan.harness_slugs) or "none selected"))
    cleared, failures = harnesses.deconfigure_selected_harnesses(
        list(plan.harness_slugs), home=home, environment=environment
    )
    result.harnesses_cleared = tuple(cleared)
    result.failures.extend(failures)
    for slug, path, changed in cleared:
        on_event(
            StepEvent(
                "harnesses",
                "finished" if changed else "skipped",
                f"{slug}: {'removed from' if changed else 'no entry in'} {path}",
            )
        )
    for slug, message in failures:
        on_event(StepEvent("harnesses", "failed", f"{slug}: {message}"))

    on_event(StepEvent("skills", "started"))
    result.skills = tuple(
        harnesses.remove_skills(
            list(plan.harness_slugs),
            plan.install_directory,
            home=home,
            environment=environment,
        )
    )
    for slug, message in result.skills:
        on_event(StepEvent("skills", "finished", f"{slug}: {message}"))

    _remove_launcher(plan, result, on_event, home=home, environment=environment)
    _remove_directories(plan, result, on_event, home=home, environment=environment)
    return result


def _remove_launcher(
    plan: UninstallPlan,
    result: UninstallResult,
    on_event: Callable[[StepEvent], None],
    *,
    home: Path | None,
    environment: Mapping[str, str] | None,
) -> None:
    bin_directory = plan.bin_directory or shell_path.default_bin_directory(
        home=home, environment=environment
    )
    on_event(StepEvent("path", "started", str(bin_directory)))
    if plan.remove_launcher:
        try:
            removed = shell_path.remove_launcher(bin_directory, plan.install_directory)
        except OSError as exc:
            result.failures.append(("launcher", str(exc)))
            on_event(StepEvent("path", "failed", str(exc)))
        else:
            result.launcher_removed = removed
            on_event(
                StepEvent(
                    "path",
                    "finished" if removed else "skipped",
                    f"removed {removed}" if removed else f"no launcher of ours in {bin_directory}",
                )
            )
    if not plan.remove_path_block:
        return
    cleared: list[Path] = []
    for profile in shell_path.shell_profiles(home=home, environment=environment):
        try:
            if shell_path.remove_path_block(profile):
                cleared.append(profile)
                on_event(StepEvent("path", "finished", f"removed the PATH block from {profile}"))
        except OSError as exc:
            result.failures.append((str(profile), str(exc)))
            on_event(StepEvent("path", "failed", f"{profile}: {exc}"))
    result.profiles_cleared = tuple(cleared)


def _remove_directories(
    plan: UninstallPlan,
    result: UninstallResult,
    on_event: Callable[[StepEvent], None],
    *,
    home: Path | None,
    environment: Mapping[str, str] | None,
) -> None:
    targets: list[tuple[Path, bool]] = []
    if plan.remove_data:
        for directory in data_directories(environment=environment):
            targets.append((directory, False))
    if plan.remove_checkout:
        targets.append((plan.install_directory, True))
    if not targets:
        return
    on_event(StepEvent("directories", "started"))
    removed: list[Path] = []
    for directory, is_checkout in targets:
        if not directory.is_dir():
            on_event(StepEvent("directories", "skipped", f"{directory} does not exist"))
            continue
        refusal = _refuse_reason(directory, checkout=is_checkout, home=home)
        if refusal is not None:
            result.failures.append((str(directory), f"not removed: it {refusal}"))
            on_event(StepEvent("directories", "failed", f"{directory} {refusal}; left alone"))
            continue
        try:
            shutil.rmtree(directory)
        except OSError as exc:
            result.failures.append((str(directory), str(exc)))
            on_event(StepEvent("directories", "failed", f"{directory}: {exc}"))
        else:
            removed.append(directory)
            on_event(StepEvent("directories", "finished", f"removed {directory}"))
    result.directories_removed = tuple(removed)


def resolve_slugs(
    selection: str | None,
    *,
    home: Path | None = None,
    environment: Mapping[str, str] | None = None,
) -> list[str]:
    """Which harnesses to clear: the named ones, or every one still configured."""

    if selection is not None:
        return harnesses.parse_harness_selection(selection)
    from .wizard import load_prefill

    return list(load_prefill(home=home, environment=environment).configured_slugs)


def describe_plan(
    plan: UninstallPlan,
    *,
    home: Path | None = None,
    environment: Mapping[str, str] | None = None,
) -> list[str]:
    """What the uninstall will do, for the confirmation prompt."""

    lines: list[str] = []
    if plan.harness_slugs:
        lines.append("Remove the code-indexing-mcp entry from:")
        for slug in plan.harness_slugs:
            lines.append(
                f"  {slug}: "
                f"{harnesses.configuration_path(slug, home=home, environment=environment)}"
            )
        lines.append("Unlink the bundled skills from those harnesses.")
    else:
        lines.append("No configured harnesses to clear.")
    bin_directory = plan.bin_directory or shell_path.default_bin_directory(
        home=home, environment=environment
    )
    if plan.remove_launcher:
        lines.append(f"Remove the launcher at {shell_path.launcher_path(bin_directory)}.")
    if plan.remove_path_block:
        lines.append("Remove the PATH block from your shell profiles.")
    if plan.remove_data:
        for directory in data_directories(environment=environment):
            lines.append(f"DELETE the index/cache directory {directory}.")
    if plan.remove_checkout:
        lines.append(f"DELETE the installation checkout {plan.install_directory}.")
    return lines


def uninstall_main(
    *,
    install_dir: str | None = None,
    harnesses_selection: str | None = None,
    bin_dir: str | None = None,
    keep_launcher: bool = False,
    keep_path: bool = False,
    purge: bool = False,
    remove_checkout: bool = False,
    assume_yes: bool = False,
    input_fn: Callable[[str], str] = input,
    output: Callable[[str], None] = print,
    error_output: Callable[[str], None] = print,
) -> int:
    """Entry for ``code-indexing-mcp uninstall``."""

    install_directory = (
        Path(install_dir).expanduser().resolve() if install_dir else default_install_directory()
    )
    try:
        slugs = resolve_slugs(harnesses_selection)
    except InstallerError as exc:
        error_output(f"Error: {exc}")
        return 1
    plan = UninstallPlan(
        install_directory=install_directory,
        harness_slugs=tuple(slugs),
        bin_directory=Path(bin_dir).expanduser() if bin_dir else None,
        remove_launcher=not keep_launcher,
        remove_path_block=not keep_path,
        remove_data=purge,
        remove_checkout=remove_checkout,
    )
    for line in describe_plan(plan):
        output(line)
    if not assume_yes:
        answer = input_fn("Proceed? [y/N]: ").strip().lower()
        if answer not in {"y", "yes"}:
            output("Uninstall cancelled.")
            return 130

    def report(event: StepEvent) -> None:
        line = f"[{event.step}] {event.status}: {event.detail}"
        (error_output if event.status == "failed" else output)(line)

    result = run_uninstall(plan, report)
    if result.failures:
        error_output(f"Uninstall finished with {len(result.failures)} problem(s); see above.")
        return 1
    output("Uninstall complete.")
    if not plan.remove_data:
        output("Indexes and caches were kept; re-run with --purge to delete them.")
    return 0

File: ts/packages/installer/src/uninstall.ts
import fs from "node:fs";
import os from "node:os";
import path from "node:path";
import { expandUser, resolveExisting } from "./config-files.ts";
import {
  configurationPath,
  deconfigureSelectedHarnesses,
  parseHarnessSelection,
  removeSkills,
} from "./harnesses.ts";
import { isUnder } from "./links.ts";
import { defaultInstallDirectory, type StepEvent } from "./orchestrator.ts";
import { BY_NAME, defaultValue } from "./settings-spec.ts";
import {
  defaultBinDirectory,
  launcherPath,
  removeLauncher,
  removePathBlock,
  shellProfiles,
} from "./shell-path.ts";
import { loadPrefill } from "./wizard.ts";

export interface UninstallPlan {
  readonly installDirectory: string;
  readonly harnessSlugs: readonly string[];
  readonly binDirectory: string | null;
  readonly removeLauncher: boolean;
  readonly removePathBlock: boolean;
  readonly removeData: boolean;
  readonly removeCheckout: boolean;
}

export interface UninstallResult {
  harnessesCleared: readonly [string, string, boolean][];
  skills: readonly [string, string][];
  launcherRemoved: string | null;
  profilesCleared: readonly string[];
  directoriesRemoved: readonly string[];
  failures: [string, string][];
}

export function dataDirectories(environment: NodeJS.ProcessEnv = process.env): string[] {
  const directories: string[] = [];
  for (const name of ["CODE_INDEXING_DATA_DIR", "CODE_INDEXING_CACHE_DIR"] as const) {
    const setting = BY_NAME[name];
    if (setting === undefined) continue;
    const configured = environment[name] || defaultValue(setting);
    if (configured !== "") directories.push(expandUser(configured));
  }
  return directories;
}

const DATA_MARKERS = [
  "lancedb",
  "locks",
  "staging",
  "accelerator.json",
  "daemon.token",
  "daemon.log",
  "models",
];

function refuseReason(
  directory: string,
  options: { checkout: boolean; home?: string },
): string | null {
  const home = path.resolve(options.home ?? os.homedir());
  let resolved: string;
  try {
    resolved = fs.realpathSync(directory);
  } catch (error) {
    return `cannot be resolved (${error instanceof Error ? error.message : String(error)})`;
  }
  if (path.dirname(resolved) === resolved) return "is a filesystem root";
  if (resolved === home) return "is your home directory";
  if (isUnder(home, resolved)) return "contains your home directory";
  if (options.checkout) {
    if (!fs.existsSync(path.join(resolved, "pyproject.toml"))) {
      return "does not look like a code-indexing-mcp checkout (no pyproject.toml)";
    }
    if (
      !fs.existsSync(path.join(resolved, "src", "code_indexing_mcp")) &&
      !fs.existsSync(path.join(resolved, "ts", "packages", "server"))
    ) {
      return "does not look like a code-indexing-mcp checkout (no src/code_indexing_mcp)";
    }
    return null;
  }
  if (path.basename(resolved) === "code-indexing-mcp") return null;
  if (DATA_MARKERS.some((marker) => fs.existsSync(path.join(resolved, marker)))) return null;
  return "holds no code-indexing-mcp index or cache, so it is not ours to delete";
}

export function runUninstall(
  plan: UninstallPlan,
  onEvent: (event: StepEvent) => void = () => undefined,
  options: { home?: string; environment?: NodeJS.ProcessEnv } = {},
): UninstallResult {
  const result: UninstallResult = {
    harnessesCleared: [],
    skills: [],
    launcherRemoved: null,
    profilesCleared: [],
    directoriesRemoved: [],
    failures: [],
  };

  onEvent({
    step: "harnesses",
    status: "started",
    detail: plan.harnessSlugs.join(", ") || "none selected",
  });
  const [cleared, failures] = deconfigureSelectedHarnesses(plan.harnessSlugs, options);
  result.harnessesCleared = cleared;
  result.failures.push(...failures);
  for (const [slug, filePath, changed] of cleared) {
    onEvent({
      step: "harnesses",
      status: changed ? "finished" : "skipped",
      detail: `${slug}: ${changed ? "removed from" : "no entry in"} ${filePath}`,
    });
  }
  for (const [slug, message] of failures) {
    onEvent({ step: "harnesses", status: "failed", detail: `${slug}: ${message}` });
  }

  onEvent({ step: "skills", status: "started", detail: "" });
  result.skills = removeSkills(plan.harnessSlugs, plan.installDirectory, options);
  for (const [slug, message] of result.skills) {
    onEvent({ step: "skills", status: "finished", detail: `${slug}: ${message}` });
  }

  removeLauncherStep(plan, result, onEvent, options);
  removeDirectories(plan, result, onEvent, options);
  return result;
}

function removeLauncherStep(
  plan: UninstallPlan,
  result: UninstallResult,
  onEvent: (event: StepEvent) => void,
  options: { home?: string; environment?: NodeJS.ProcessEnv },
): void {
  const binDirectory = plan.binDirectory ?? defaultBinDirectory(options);
  onEvent({ step: "path", status: "started", detail: binDirectory });
  if (plan.removeLauncher) {
    try {
      const removed = removeLauncher(binDirectory, plan.installDirectory);
      result.launcherRemoved = removed;
      onEvent({
        step: "path",
        status: removed !== null ? "finished" : "skipped",
        detail: removed !== null ? `removed ${removed}` : `no launcher of ours in ${binDirectory}`,
      });
    } catch (error) {
      result.failures.push(["launcher", error instanceof Error ? error.message : String(error)]);
      onEvent({
        step: "path",
        status: "failed",
        detail: error instanceof Error ? error.message : String(error),
      });
    }
  }
  if (!plan.removePathBlock) return;
  const cleared: string[] = [];
  for (const profile of shellProfiles(options)) {
    try {
      if (removePathBlock(profile)) {
        cleared.push(profile);
        onEvent({
          step: "path",
          status: "finished",
          detail: `removed the PATH block from ${profile}`,
        });
      }
    } catch (error) {
      result.failures.push([profile, error instanceof Error ? error.message : String(error)]);
      onEvent({
        step: "path",
        status: "failed",
        detail: `${profile}: ${error instanceof Error ? error.message : String(error)}`,
      });
    }
  }
  result.profilesCleared = cleared;
}

function removeDirectories(
  plan: UninstallPlan,
  result: UninstallResult,
  onEvent: (event: StepEvent) => void,
  options: { home?: string; environment?: NodeJS.ProcessEnv },
): void {
  const targets: [string, boolean][] = [];
  if (plan.removeData) {
    for (const directory of dataDirectories(options.environment)) targets.push([directory, false]);
  }
  if (plan.removeCheckout) targets.push([plan.installDirectory, true]);
  if (targets.length === 0) return;
  onEvent({ step: "directories", status: "started", detail: "" });
  const removed: string[] = [];
  for (const [directory, isCheckout] of targets) {
    if (!fs.existsSync(directory) || !fs.statSync(directory).isDirectory()) {
      onEvent({ step: "directories", status: "skipped", detail: `${directory} does not exist` });
      continue;
    }
    const refusal = refuseReason(directory, {
      checkout: isCheckout,
      ...(options.home === undefined ? {} : { home: options.home }),
    });
    if (refusal !== null) {
      result.failures.push([directory, `not removed: it ${refusal}`]);
      onEvent({
        step: "directories",
        status: "failed",
        detail: `${directory} ${refusal}; left alone`,
      });
      continue;
    }
    try {
      fs.rmSync(directory, { recursive: true, force: false });
      removed.push(directory);
      onEvent({ step: "directories", status: "finished", detail: `removed ${directory}` });
    } catch (error) {
      result.failures.push([directory, error instanceof Error ? error.message : String(error)]);
      onEvent({
        step: "directories",
        status: "failed",
        detail: `${directory}: ${error instanceof Error ? error.message : String(error)}`,
      });
    }
  }
  result.directoriesRemoved = removed;
}

export function resolveSlugs(
  selection: string | null,
  options: { home?: string; environment?: NodeJS.ProcessEnv } = {},
): string[] {
  if (selection !== null) return parseHarnessSelection(selection);
  return [...loadPrefill(options).configuredSlugs];
}

export function describePlan(
  plan: UninstallPlan,
  options: { home?: string; environment?: NodeJS.ProcessEnv } = {},
): string[] {
  const lines: string[] = [];
  if (plan.harnessSlugs.length > 0) {
    lines.push("Remove the code-indexing-mcp entry from:");
    for (const slug of plan.harnessSlugs) {
      lines.push(`  ${slug}: ${configurationPath(slug, options)}`);
    }
    lines.push("Unlink the bundled skills from those harnesses.");
  } else {
    lines.push("No configured harnesses to clear.");
  }
  const binDirectory = plan.binDirectory ?? defaultBinDirectory(options);
  if (plan.removeLauncher) {
    lines.push(`Remove the launcher at ${launcherPath(binDirectory)}.`);
  }
  if (plan.removePathBlock) lines.push("Remove the PATH block from your shell profiles.");
  if (plan.removeData) {
    for (const directory of dataDirectories(options.environment)) {
      lines.push(`DELETE the index/cache directory ${directory}.`);
    }
  }
  if (plan.removeCheckout) {
    lines.push(`DELETE the installation checkout ${plan.installDirectory}.`);
  }
  return lines;
}

export async function uninstallMain(options: {
  installDir?: string | null;
  harnessesSelection?: string | null;
  binDir?: string | null;
  keepLauncher?: boolean;
  keepPath?: boolean;
  purge?: boolean;
  removeCheckout?: boolean;
  assumeYes?: boolean;
  inputFn?: (prompt: string) => Promise<string> | string;
  output?: (line: string) => void;
  errorOutput?: (line: string) => void;
}): Promise<number> {
  const output = options.output ?? console.log;
  const errorOutput = options.errorOutput ?? console.error;
  const installDirectory =
    options.installDir != null ? resolveExisting(options.installDir) : defaultInstallDirectory();
  let slugs: string[];
  try {
    slugs = resolveSlugs(options.harnessesSelection ?? null);
  } catch (error) {
    errorOutput(`Error: ${error instanceof Error ? error.message : String(error)}`);
    return 1;
  }
  const plan: UninstallPlan = {
    installDirectory,
    harnessSlugs: slugs,
    binDirectory: options.binDir == null ? null : expandUser(options.binDir),
    removeLauncher: options.keepLauncher !== true,
    removePathBlock: options.keepPath !== true,
    removeData: options.purge === true,
    removeCheckout: options.removeCheckout === true,
  };
  for (const line of describePlan(plan)) output(line);
  if (options.assumeYes !== true) {
    const ask = options.inputFn ?? ((prompt: string) => prompt);
    const answer = String(await ask("Proceed? [y/N]: "))
      .trim()
      .toLowerCase();
    if (answer !== "y" && answer !== "yes") {
      output("Uninstall cancelled.");
      return 130;
    }
  }
  const result = runUninstall(plan, (event) => {
    const line = `[${event.step}] ${event.status}: ${event.detail}`;
    (event.status === "failed" ? errorOutput : output)(line);
  });
  if (result.failures.length > 0) {
    errorOutput(`Uninstall finished with ${result.failures.length} problem(s); see above.`);
    return 1;
  }
  output("Uninstall complete.");
  if (!plan.removeData) {
    output("Indexes and caches were kept; re-run with --purge to delete them.");
  }
  return 0;
}
